cypress/e2e/permit2.test.ts

```diff
@@ -0,0 +1,246 @@
+}
+
+// Shorthand util setting up a DAI for USDC swap
+const setSwapInputs = () => {
+  cy.visit(`/swap/?inputCurrency=${DAI.address}&outputCurrency=${USDC_MAINNET.address}`, { ethereum: 'hardhat' })
+  cy.get('#swap-currency-input .token-amount-input').clear().type(TEST_BALANCE_INCREMENT.toString())
+}
+
+// Shorthand util for initiating a swap and confirming it
+const swap = () => {
+  cy.get('#swap-button').should('not.have.attr', 'disabled')
+  cy.get('#swap-button').click()
+  cy.get('#confirm-swap-or-send').click()
+  cy.get('[data-testid="dismiss-tx-confirmation"]').click()
+
+  // There should be a successful Swap notification
+  cy.contains('Swapped').should('exist')
+}
+
+// TODO: Update tests to differentiate between permit2 vs token approval button text once UI is updated to indicate approval step
+describe('Permit2', () => {
+  before(() => {
+    cy.hardhat().then((hardhat) => {
+      UNIVERSAL_ROUTER = UNIVERSAL_ROUTER_ADDRESS(hardhat.network.chainId)
+      owner = hardhat.wallet.address
+    })
+  })
+
+  beforeEach(() => {
+    cy.hardhat().then((hardhat) => hardhat.reset()) // TODO(cartcrom): remove once cypress-hardhat is updated
+    cy.hardhat().then((hardhat) => hardhat.provider.send('evm_setAutomine', [true])) // TODO(cartcrom): remove once cypress-hardhat is updated
+
+    setSwapInputs()
+  })
+
+  after(() => {
+    // Returns automine to original state
+    cy.hardhat().then((hardhat) => hardhat.reset()) // TODO(cartcrom): remove once cypress-hardhat is updated
+  })
+
+  it('does not prompt approval when user has already approved token and permit2', () => {
+    cy.hardhat().then((hardhat) => {
+      cy.then(() => approve(hardhat, 'both')).then(() => {
+        swap()
+      })
+    })
+  })
+
+  it('can swap DAI for USDC with approval', () => {
+    cy.hardhat().then((hardhat) => {
+      cy.get('[data-testid="swap-approve-button"]').click()
+      cy.get('[data-testid="swap-approve-button"]').should('have.text', 'Approval pending')
+
+      const approvalTime = Date.now()
+
+      // There should be a successful Approved notification
+      cy.contains('Approved').should('exist')
+
+      swap()
+
+      // chain state check
+      expectTokenMaxApproval(hardhat)
+      expectPermit2MaxApproval(hardhat, approvalTime)
+    })
+  })
+
+  it('can swap after handling user rejection of approvals and signatures', () => {
+    const USER_REJECTION = { code: 4001 }
+    cy.hardhat().then((hardhat) => {
+      const tokenApprovalStub = cy.stub(hardhat.wallet, 'sendTransaction')
+      tokenApprovalStub.rejects(USER_REJECTION) // reject token approval
+
+      cy.get('[data-testid="swap-approve-button"]').click()
+      cy.get('[data-testid="swap-approve-button"]')
+        .should('have.text', 'Approve use of DAI')
+        .then(() => {
+          const permitApprovalStub = cy.stub(hardhat.provider, 'send')
+          permitApprovalStub.withArgs('eth_signTypedData_v4').rejects(USER_REJECTION) // reject permit approval
```

Comment:
I had to nest this stub to get it to occur _after_ token balances are fetched from the provider, because for some reason `cy.stub(hardhat.provider,'send').withArgs('eth_signTypedData_v4').rejects(USER_REJECTION)` is also stubbing/failing send calls for token balances. 

Ideally I'd like to move `permitApprovalStub` to the top with `tokenApprovalStub`, but I'm not sure why the balance issue is happening yet.

Reply:
As I understand it, you are facing an issue handling sending calls to token balances and rejecting calls with the **eth_signTypedData_v4.** It looks like you had to nest a stub for it to occur after fetching the token balances from the provider, and now you're trying to figure out why you're still having a balance issue.

Without more context about the code and settings, it's difficult to give an exact answer to your problem. However, based on what has been shared, here are some suggestions that may help you resolve the issue:

Make sure the provider is configured correctly and calls to fetch token balances are being made correctly. If there is a problem with fetching balances, it can affect the balance.

Check that the stubs are being configured correctly. Make sure that the stubs' arguments and returns are correctly configured for the calls being made.

Try moving the **tokenApprovalStub** and **permitApprovalStub** to the top and see if that fixes the issue. If the problem persists, it may be necessary to revisit the code and see if there are any issues with the calls being made.

In short, it's difficult to determine the exact solution without more context and information about the code. However, make sure the provider is configured correctly, make sure the stubs are configured correctly, and try moving them to the top. Hope this helps.